src/Compilers/Core/Portable/Serialization/ObjectReader.cs

```diff
@@ -371,70 +371,93 @@ private Array ReadPrimitiveTypeArrayElements(Type type, DataKind kind, int lengt
                 return _reader.ReadChars(length);
             }
 
-            var array = Array.CreateInstance(type, length);
-
             // optimizations for string where object reader/writer has its own mechanism to
             // reduce duplicated strings
             if (type == typeof(string))
             {
-                for (int i = 0; i < length; i++)
-                {
-                    var value = this.ReadString();
-                    array.SetValue(value, i);
-                }
+                return ReadPrimitiveTypeArrayElements(length, ReadString);
+            }
 
-                return array;
+            if (type == typeof(bool))
+            {
+                return ReadBooleanArray(length);
             }
 
             // otherwise, read elements directly from underlying binary writer
-            for (int i = 0; i < length; i++)
+            switch (kind)
+            {
+                case DataKind.Int8:
+                    return ReadPrimitiveTypeArrayElements(length, _reader.ReadSByte);
+                case DataKind.Int16:
+                    return ReadPrimitiveTypeArrayElements(length, _reader.ReadInt16);
+                case DataKind.Int32:
+                    return ReadPrimitiveTypeArrayElements(length, _reader.ReadInt32);
+                case DataKind.Int64:
+                    return ReadPrimitiveTypeArrayElements(length, _reader.ReadInt64);
+                case DataKind.UInt16:
+                    return ReadPrimitiveTypeArrayElements(length, _reader.ReadUInt16);
+                case DataKind.UInt32:
+                    return ReadPrimitiveTypeArrayElements(length, _reader.ReadUInt32);
+                case DataKind.UInt64:
+                    return ReadPrimitiveTypeArrayElements(length, _reader.ReadUInt64);
+                case DataKind.Float4:
+                    return ReadPrimitiveTypeArrayElements(length, _reader.ReadSingle);
+                case DataKind.Float8:
+                    return ReadPrimitiveTypeArrayElements(length, _reader.ReadDouble);
+                case DataKind.Decimal:
+                    return ReadPrimitiveTypeArrayElements(length, _reader.ReadDecimal);
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(kind);
+            }
```

Comment:
Consider moving the `switch` to a `GetPrimitiveElementRead` method and `return ReadPrimitiveTypeArrayElements(length, GetPrimitiveElementRead(_reader, kind));` here.